feat(data_source): OSV

rhythmrx9 opened the pull request:
closes #1749 
added `osv_source` as a new `data_source`.

terriko commented:
Notes for me: adds two new libraries

* cvss is LGPLv3+ and comes from someone at redhat, so probably can get approval https://pypi.org/project/cvss/
* gsutil is Apache and comes from Google, so probably also like to get approval https://pypi.org/project/gsutil/

terriko commented:
More general comments: I like the ease of sql's "INSERT OR REPLACE" and "INSERT OR IGNORE" to merge the data without duplicates, but I'm wondering if users might need/want more customizability than just "replace for highs, don't replace for lows"

Possible options we might offer:
1. As written, replace NVD rating for highs, do not for lows
2. Always prefer NVD
3. Always prefer OSV
4. Let users set their own threshold at which point they prefer one over the other.
5. Prefer whichever rating is higher
6. Prefer whichever rating is lower
7. Prefer whichever rating is most recent
8. Somehow store and display both in reports so users could make decisions (makes interface messy and confusing, pushes decision on end users)
9. Choose and display only one but put a note/comment/special field indicating "source X has a different rating [link]" or "source X has the same rating"

We'll probably need to think about how to handle this for arbitrary numbers of different sources too.  Priority list for all? Threshold list for all?

terriko commented:
(updating branch for CI)

terriko commented:
Okay, I think those test failures may actually be real?  At least, I'd expect the short tests to pass completely on PRs and they're passing on main at the moment.

anthonyharrison reviewed (commented):
In the osv_source, CVEs without a CVSS V3 score are ignored. Suggest that if a CVSS 3 vector is not found that a check for a CVSS V2 vector is performed.

Would be good if some comments were added to the  file to understand the flow of logic

rhythmrx9 commented:
@anthonyharrison  OSV schema only provides CVSS V3 scores as of now, the check is there for if CVSS V2 scores or any other scores are added to it in the future. 
You're right, I should add some comments for that, thanks.

rhythmrx9 commented:
@terriko,  added `-d --disable-data-source` option to disable a data source, and documentation is also updated.
Will open a separate PR for `OSV` tests.

terriko commented:
this is looking good, but I'm still getting some test failures.  I've kicked off those to re-run.  Normally I'd want the tests directly in the same PR but given how big this is getting I think you're right that it could be separate.

terriko commented:
I did not expect this to have a conflict with the spelling PR, sorry.  I think it's resolved.  Tests were passing before I had to do that, we'll probably merge this code tomorrow since I need to go eat dinner.

terriko reviewed (approved):
Filed an issue about those failing tests but didn't encounter the same problem when I re-ran them.

I'm going to go ahead and merge this now so you can focus on the tests next.
